### ISSUE TITLE
Add plugin to support passing .NET objects when using Remote/Cluster actors

### ISSUE THREAD

**Issue by raskolnikoov** [AsynkronIT/protoactor-dotnet#248]
When using remote/cluster actors today between .NET actors we cannot pass .NET objects, it must be a protobuf message. In some usecases some might want to pass .NET objects between those actors that will always be running on .NET. So a plugin to allow the actors to support this would be nice to have.

Maybe this plugin could be something to write support for first? ;)

https://github.com/rogeralsing/Wire

**Comment by cpx86** [comment]
Something that could be quite cool is if we could have content negotiation between remote processes. It would be a pretty major change to Remote though, so we should think about it carefully.

I'm thinking something like this:
1. Process A connects to Process B
2. Process B sends list of supported/preferred MIME types, e.g. `application/x-protobuf,application/json,application/x-wiredotnet`
3. Process A takes the first supported MIME type and configures itself to use it
4. Process A sends the decided MIME type to Process B
5. Process A and B will use the associated serializer when serializing/deserializing message envelopes

This would also require a big change to how we register known message types. Currently we use Protobuf file descriptors - it would have to be made a lot more generic to work with other serializers. I'm also not sure how to construct the mapping between message names and message types in a good way.

Maybe this is too complicated and isn't a good idea after all. I don't know, it's just an idea :) What are your thoughts? Ping @rogeralsing

**Comment by raskolnikoov** [comment]
@cpx86 I dont think its a bad idea at all. I like it alot! Sure maybe performance could downgrade but it would give it alot more usecases. :)

**Comment by rogeralsing** [comment]
After AkkaNET I'd like to stay as far away as possible from overly complicated serialization.
In .NET we can probably do a lot of fancy things.

But remember that PID is a protobuf type, and making that play nice with Avro or Thrift or whatever would probably be weird.

I would however want to be able to pass JSON payloads to make it easier to support the CLI tool.
And this could probably be extended to support other formats.
I'm thinking that maybe the remote.MessageEnvelope could be extended with a serializer ID.
That way, 0 could be the default protobuf serializer, 1 could be protobuf compatible JSON and N could be a custom serializer.

e.g we could allow for Wire to be used for .NET only systems.

The Akka way is to associate a specific serializer per message type, I'd like to avoid that, I find it overly complex and people tend to not get the problems this results in.
e.g. type 1 uses serilizer x, then type 2 using serializer y while also having a property of type 1.
in that case, the nested type 1 in type 2 would be serialized with serializer y.
MS Orleans will try to serialize each sub field using the serializer for that type.

TLDR; it's messed up, complicated, bug prone, slow, and not the way I'd like to go.

So IMO, lets just use one and the same serializer for your entire system, default to ProtoBuf, but make it easy to activate Wire which then replaces protobuf completely.

**Comment by cpx86** [comment]
@rogeralsing I agree the Akka way as you describe it is way too complicated. I think we'd avoid most of those issues though if we were to do an initial handshake between the processes, to agree on which format all messages should be serialized in. Internally, it would require some changes, but I think we could do it in a way that's not very complicated to the user.

**Comment by rogeralsing** [comment]
Solved by #261
